### ISSUE TITLE
Time intervals/deltas and transformations.

### ISSUE THREAD

**Issue by frnco** [mint-lang/mint#521]
While working with data that expires I notices it's quite hard to get a `Time` that is relative to `Time.now`, and after searching the issues I wasn't able to find any proposals regarding this, except for [137](https://github.com/mint-lang/mint/issues/137) which actually gave me some ideas on a simple solution to make this work.

My proposal is fairly simple, but first I'll list some options I've considered. The first was adding stuff inline with `nextWeek`, stuff like `tomorrow` etc, but that's still very limited so I went on to consider adding Arithmethical/Transformation functions, i.e. `Time.now |> Time.addDays(2)` or `Time.now |> Time.addMinutes(30)`. But since we already have Unix, it seems pretty easy to just have something like `(Time.toUnixTimestamp(TIme.now) + ( 30 * 60 * 1000 )) |> Time.fromUnixTimestamp`, here considering the `UnixTimestamp` thing in Mint actually takes milliseconds into account, though I believe it'd be nice to include some functions into `Time` to allow us to explicitly make that distinction.

The change would actually be limited to the `TIme` module, and would consist of simply adding `toSomeEpochFormat` and `fromSomeEpochFormat` functions, which would allow anyone to modify time using simple arithmetic.

Sure, it'd be cool to have something like Ruby available, i.e. `TIme.now |> Time.add(Time.days(3))` but unfortunately `Time` does not deal with deltas. I did start to work on a `Timespan` module to deal with that but I'm nowhere near happy with how I went about doing it (A big factor was my failure to figure out how to make it a custom type consisting of basically a single number, which corresponds to the number of milliseconds for that timespan/timedelta. Instead I had to use Javascript to convert it to and from `Number`, which completely screwed up the elegant approach I came up with at the beginning and made me put that idea aside, at least for now). I wanna get back to that but first I need to figure out a way to make it a custom type that I can do arithmetic with and convert to and from Numbers without having to use inline Javascript, but the only way I can see that happening means changing the compiler, so it's postponed until I figured out a way to deal with the compiler (Or how to make the compiler friendlier), unfortunately.

Any suggestions on how to make this easier are appreciated, but for now the easiest option I can see is changing the `Time` module to painlessly convert to and from UNIX/POSIX Timestamps/Epochs, which would allow anyone to use `Number`s' to do basic math to modify the time and then convert it back. I can make the change myself and send a PR with the modified `core/source/Time.mint`, basically it'd be something like:
```
fun fromUnixTimestamp (timestamp : Number) : Time {
  `
  (() => {
    const x = new Date()
    x.setTime(#{timestamp})
    return x
  })()
  `
}
fun toUnixTimestamp (time : Time) : Number {
  `
  (() => {
    const x = new Date(+#{time})
    return x.getTime()
  })()
  `
}
```
Plus the respective tests in `core/tests/tests/Time.mint`, in the tests I used the standard `April 5th, 2018` which all others tests use, so it matches the other tests, and I checked the timestamp:
```
suite "Time.fromUnixTimestampInMs" {
  test "returns the time corresponding to Unix Timestamp in Milliseconds" {
    (Time.fromUnixTimestamp(1522886400000)
    |> Time.getTime()) == "2018-04-05T00:00:00.000Z"
  }
}
suite "Time.toUnixTimestampInMs" {
  test "returns the Unix Timestamp in Milliseconds corresponding to time" {
    (Time.from(2018, 4, 5)
    |> Time.toUnixTimestamp()) == 1522886400000
  }
}
```

I'd love to know whether this is acceptable or not, and whether sending PR would help, and I'd love to build on it if anyone has something else to suggest.

**Comment by gdotdesign** [comment]
I think `Time.fromUnixTimestampInMs` and `Time.toUnixTimestampInMs` would be good for the current state, the only change I would make is to have all dates in UTC, easier to reason about and only use time zones for display purposes (at least for me).

--------------------------------------

Thanks for looking into the `Time` module, by far it's the most underdeveloped module in the standard library, and it needs a big makeover. Its current state is basically some functions with inline JavaScript and some functions from DateFNS which are bundled with the runtime.

If you are willing to tackle it, I would be happy to brainstorm about the overall API.

**Comment by frnco** [comment]
> Thanks for looking into the `Time` module, by far it's the most underdeveloped module in the standard library, and it needs a big makeover. Its current state is basically some functions with inline JavaScript and some functions from DateFNS which are bundled with the runtime.

The biggest issue with it right now, as far as I can see, is that despite it being called `Time` it's actually based on the `Date` module from JS, making it hard to find a decent name for a complementary counterpart module that deals with time intervals. Ideally I'd prefer having `Time` deal with intervals and adding something like `Date` or `Datetime` to deal with actual specific moments in time, but it is what it is, I'm currently considering `Duration`, `TimeDelta` and `TimeSpan` (Possibly shortened to `TDelta`/`TSpan`) to complement the `Time` module. But if for any reason it's decided that the API is to be changed, I suggest thinking about that.

For now I've been working on two fronts, one is handling timespans, like `Duration.hours(2)` and `Duration.weeks(1)`, but that made it rpetty obvious that it'd be less than ideal to simply convert stuff like `Duration.months()` and `Duration.years()` to some arbitrary number of seconds/milliseconds, since their duration changes depending on which date is being modified by them and whether those are being added or removed to said date, thus I started a second approach which actually just modifies the `Time`, for now I'm using `TDelta` as the module's name and I'll include some of the code and tests below, if you have any suggestions I'd love to hear them.

That did get me thinking that maybe it'd be a good idea to add the option of having Deltas in the `Time` module itself, if it's structured like what I've been working on that'd mean adding an `Enum` to differentiate between units (Milliseconds, Seconds, Minutes, Hours, Days, Weeks, Months, Years), and a class/module/etc to deal with actually modifying the `Time`.

This is the gist of what I made:
```
enum TDelta.Unit {
  Milliseconds(Number)
  Seconds(Number)
  Minutes(Number)
  Hours(Number)
  Days(Number)
  Weeks(Number)
  Months(Number)
  Years(Number)
}
module TDelta {
  fun timeMoveMilliseconds(goForward : Bool, baseTime : Time, deltaInMilliseconds : Number) : Time {
    `(() => {
      const x = new Date(+#{baseTime})
      if (#{goForward}) {
        x.setTime(#{baseTime}.getTime() + #{deltaInMilliseconds})
      } else {
        x.setTime(#{baseTime}.getTime() - #{deltaInMilliseconds})
      }
      return x
    })()`
  }
  fun timeMoveDays(goForward : Bool, baseTime : Time, deltaInDays : Number) : Time {
    `(() => {
      const x = new Date(+#{baseTime})
      if (#{goForward}) {
        x.setDate(#{baseTime}.getDate() + #{deltaInDays})
      } else {
        x.setDate(#{baseTime}.getDate() - #{deltaInDays})
      }
      return x
    })()`
  }
  fun timeMoveMonths(goForward : Bool, baseTime : Time, deltaInMonths : Number) : Time {
    `(() => {
      const x = new Date(+#{baseTime})
      if (#{goForward}) {
        x.setMonth(#{baseTime}.getMonth() + #{deltaInMonths})
      } else {
        x.setMonth(#{baseTime}.getMonth() - #{deltaInMonths})
      }
      return x
    })()`
  }
  fun timeMoveYears(goForward : Bool, baseTime : Time, deltaInYears : Number) : Time {
    `(() => {
      const x = new Date(+#{baseTime})
      if (#{goForward}) {
        x.setFullYear(#{baseTime}.getFullYear() + #{deltaInYears})
      } else {
        x.setFullYear(#{baseTime}.getFullYear() - #{deltaInYears})
      }
      return x
    })()`
  }
  fun modifyTime(goForward : Bool, delta : TDelta.Unit, baseTime : Time) : Time {
    case (delta) {
      TDelta.Unit::Millisecond(amount) => amount |> timeMoveMilliseconds(goForward, baseTime)
      TDelta.Unit::Second(amount) => (amount * 1000) |> timeMoveMilliseconds(goForward, baseTime)
      TDelta.Unit::Minute(amount) => (amount * 1000 * 60) |> timeMoveMilliseconds(goForward, baseTime)
      TDelta.Unit::Hour(amount) => (amount * 1000 * 60 * 60) |> timeMoveMilliseconds(goForward, baseTime)
      TDelta.Unit::Day(amount) => (amount) |> timeMoveDays(goForward, baseTime)
      TDelta.Unit::Week(amount) => (amount * 7) |> timeMoveDays(goForward, baseTime)
      TDelta.Unit::Month(amount) => (amount) |> timeMoveMonths(goForward, baseTime)
      TDelta.Unit::Year(amount) => (amount) |> timeMoveYears(goForward, baseTime)
    }
  }
}
```
and the tests, which actually were all green last time I checked (I'll just paste everything, hope it doesn't get too unreasonably long):
```
suite "TDelta" {
  test "Creates Delta" {
    TDelta.new(TDelta.Unit::Millisecond, 0) == {unit=TDelta.Unit::Millisecond,amount=0}
  }
}
suite "Moves time Forward" {
  test "Moves millisecond forward" {
    (Time.from(2018, 4, 5)
    |> TDelta.timeAfter(TDelta.milliseconds(1))
    |> Time.toIso()) == "2018-04-05T00:00:00.001Z"
  }
  test "Moves second forward" {
    (Time.from(2018, 4, 5)
    |> TDelta.timeAfter(TDelta.seconds(1))
    |> Time.toIso()) == "2018-04-05T00:00:01.000Z"
  }
  test "Moves minute forward" {
    (Time.from(2018, 4, 5)
    |> TDelta.timeAfter(TDelta.minutes(1))
    |> Time.toIso()) == "2018-04-05T00:01:00.000Z"
  }
  test "Moves hour forward" {
    (Time.from(2018, 4, 5)
    |> TDelta.timeAfter(TDelta.hours(1))
    |> Time.toIso()) == "2018-04-05T01:00:00.000Z"
  }
  test "Moves day forward" {
    (Time.from(2018, 4, 5)
    |> TDelta.timeAfter(TDelta.days(1))
    |> Time.toIso()) == "2018-04-06T00:00:00.000Z"
  }
  test "Moves week forward" {
    (Time.from(2018, 4, 5)
    |> TDelta.timeAfter(TDelta.weeks(1))
    |> Time.toIso()) == "2018-04-12T00:00:00.000Z"
  }
  test "Moves month forward" {
    (Time.from(2018, 4, 5)
    |> TDelta.timeAfter(TDelta.months(1))
    |> Time.toIso()) == "2018-05-05T00:00:00.000Z"
  }
  test "Moves year forward" {
    (Time.from(2018, 4, 5)
    |> TDelta.timeAfter(TDelta.years(1))
    |> Time.toIso()) == "2019-04-05T00:00:00.000Z"
  }
}
suite "Moves time Backward" {
  test "Moves millisecond backward" {
    (Time.from(2018, 4, 5)
    |> TDelta.timeBefore(TDelta.milliseconds(1))
    |> Time.toIso()) == "2018-04-04T23:59:59.999Z"
  }
  test "Moves second backward" {
    (Time.from(2018, 4, 5)
    |> TDelta.timeBefore(TDelta.seconds(1))
    |> Time.toIso()) == "2018-04-04T23:59:59.000Z"
  }
  test "Moves minute backward" {
    (Time.from(2018, 4, 5)
    |> TDelta.timeBefore(TDelta.minutes(1))
    |> Time.toIso()) == "2018-04-04T23:59:00.000Z"
  }
  test "Moves hour backward" {
    (Time.from(2018, 4, 5)
    |> TDelta.timeBefore(TDelta.hours(1))
    |> Time.toIso()) == "2018-04-04T23:00:00.000Z"
  }
  test "Moves day backward" {
    (Time.from(2018, 4, 5)
    |> TDelta.timeBefore(TDelta.days(1))
    |> Time.toIso()) == "2018-04-04T00:00:00.000Z"
  }
  test "Moves week backward" {
    (Time.from(2018, 4, 5)
    |> TDelta.timeBefore(TDelta.weeks(1))
    |> Time.toIso()) == "2018-03-29T00:00:00.000Z"
  }
  test "Moves month backward" {
    (Time.from(2018, 4, 5)
    |> TDelta.timeBefore(TDelta.months(1))
    |> Time.toIso()) == "2018-03-05T00:00:00.000Z"
  }
  test "Moves year backward" {
    (Time.from(2018, 4, 5)
    |> TDelta.timeBefore(TDelta.years(1))
    |> Time.toIso()) == "2017-04-05T00:00:00.000Z"
  }
}
```


Making it part of Time could be interesting, it'd mean changing it from `TDelta` to `Time.Delta`, and the Units could either be `Time.Units` or `Time.Delta.Units`, either way that should allow something more readable like:
```
(Time.from(2018, 4, 5)
|> Time.deltaAfter(Time.Delta.months(2))
|> Time.toIso()) == "2018-06-05T00:00:00.000Z"
```
and/or
```
(Time.from(2018, 4, 5)
|> Time.deltaBefore(Time.Delta.months(2))
|> Time.toIso()) == "2018-02-05T00:00:00.000Z"
```
`deltaBefore` and `deltaAfter` are quite bad names, I was using `timeBefore` and `timeAfter` but that becomes `TDelta.timeBefore` and `TDelta.timeAfter` using `TDelta` as the base module, which makes some sense, but if those are on `Time` module it just looks bad, so I'm trying to decide on a better name for those methods. Either way, if you think including those in the `Time` module may be a good idea, I can also send some code for that.


In the future I'd love to also have something similar to what Rails (ActiveSupport, more specifically) does for converting time to something that makes sense in human language, like https://apidock.com/rails/ActionView/Helpers/DateHelper/distance_of_time_in_words and https://apidock.com/rails/ActionView/Helpers/DateHelper/time_ago_in_words etc., but it would probably be better to have a separate module for that, preferably as a package and not on the core.



As for `Time` being the most underdeveloped module, I don't think it's that bad, actually. `Time` works quite well, I think there are plenty other modules in the standard library that deserve more attention, not to mention creating syntax to make it easier to encode/decode data and/or add the option of dinamically generating `Enum` options at compile-time, but even if we only consider the more "simple" stuff, Mint could really use some nicer ways to reference elements in Maps/Arrays, so we could render one of them and, upon changing the values, not having to double and triple and endlessly-check what does and/or doesn't exist, cross-referencing stuff etc., just for the sake of changing one string in a list. It's pretty hard to find a generalized solution to dealing with that, sure, but I think that is definitely inconveniencing a lot more people than any perks of the `Time` module.


For now I'll be forking the repo and start writing the tests to make the `Time` module easier to work with, while I wait for your opinion on making it able to understand relative times and get relative times using `before` and `after` or whatever we end up with (Which is hopefully better than what I've used here).

**Comment by gdotdesign** [comment]
I like what you started to do here. I think we can get some ideas for our own implementation from Crystal https://crystal-lang.org/api/1.2.2/Time.html I really like their API. 

For the duration they use `Time::Span` https://crystal-lang.org/api/1.2.2/Time/Span.html which I like more than `Delta` or `Duration` so I would go with:

```
enum Time.Span {
  Milliseconds(Number)
  Seconds(Number)
  Minutes(Number)
  Hours(Number)
  Days(Number)
  Weeks(Number)
  Months(Number)
  Years(Number)
}
```

I like what you have started with the `shift` method in the other issue, and I've built upon it to explore this API here https://sandbox.mint-lang.com/sandboxes/ebAXP5g3Sj2wUA

- instead of `direction` we can just use negative numbers to go backwards - `Time.Span::Days(-1)` 
- all functions _clone_ the date to return a new instance (immutable)
- all transformations are done on the UTC version of the date

**Comment by gdotdesign** [comment]
Also, here is what I would like to replace with our implementation from the existing functions to drop dependency on `DateFNS`:

- `Time.format`, `Time.relative` - we need to look at other implementations (maybe https://package.elm-lang.org/packages/ryannhg/date-format/latest/)
- `Time.startOf` - `Time.atBeginningOfDay`, `Time.atBeginningOfMonth`, etc...
- `Time.endOf` - `Time.atEndOfDay`, `Time.atEndOfMonth`, etc...
- `Time.range` - We can copy from https://github.com/date-fns/date-fns/blob/4544f0cb1d724bfa7ee846b762bacb8fc303a961/src/eachDayOfInterval/index.ts

I got a little excited about this, so I'll probably do some exploration in a branch.

**Comment by frnco** [comment]
> * instead of `direction` we can just use negative numbers to go backwards - `Time.Span::Days(-1)`
> * all functions _clone_ the date to return a new instance (immutable)
> * all transformations are done on the UTC version of the date

I did realize that and already changed it.

However, to that with reasonably clear and concise code I changed the `enum` so it's used only for the units, and opted for using a Tuple (May change to a record, still thinking about it), so that it's easier to change the actual number without having to go through a `case` or something like that. I also remove `weeks` in favor of having the `Time.weeks` method multiply the number of weeks it receives by 7 and return it as `Time.Span::Days`. Here's an overview of the current state of things (Omitting the unmodified stuff in the `Time` module):
```
enum Time.Span {
  Milliseconds
  Seconds
  Minutes
  Hours
  Days
  Months
  Years
}
module Time {
  /* returns a Timedelta in Milliseconds */
  fun milliseconds(amount : Number) : Tuple(Number,Time.Span) {
    {amount, Time.Span::Milliseconds}
  }

  /* returns a Timedelta in Seconds */
  fun seconds(amount : Number) : Tuple(Number,Time.Span) {
    {amount, Time.Span::Seconds}
  }

  /* returns a Timedelta in Minutes*/
  fun minutes(amount : Number) : Tuple(Number,Time.Span) {
    {amount, Time.Span::Minutes}
  }

  /* returns a Timedelta in Hours */
  fun hours(amount : Number) : Tuple(Number,Time.Span) {
    {amount, Time.Span::Hours}
  }

  /* returns a Timedelta in Days */
  fun days(amount : Number) : Tuple(Number,Time.Span) {
    {amount, Time.Span::Days}
  }

  /* returns a Timedelta in Weeks */
  fun weeks(amount : Number) : Tuple(Number,Time.Span) {
    {(amount * 7), Time.Span::Days}
  }

  /* returns a Timedelta in Months */
  fun months(amount : Number) : Tuple(Number,Time.Span) {
    {amount, Time.Span::Months}
  }

  /* returns a Timedelta in Years */
  fun years(amount : Number) : Tuple(Number,Time.Span) {
    {amount, Time.Span::Years}
  }

  /* Shifts the time passed forward if direction is true, backward otherwise, and returns the result  */
  fun shift(offset : Number, unit : Time.Span, time : Time) : Time {
    `
    (() => {
      const time = new Date(+#{time});
      #{
        case (unit) {
          Time.Span::Milliseconds => `time.setUTCMilliseconds(time.getUTCMilliseconds() + #{offset})`
          Time.Span::Seconds      => `time.setUTCSeconds(time.getUTCSeconds() + #{offset})`
          Time.Span::Minutes      => `time.setUTCMinutes(time.getUTCMinutes() + #{offset})`
          Time.Span::Hours        => `time.setUTCHours(time.getUTCHours() + #{offset})`
          Time.Span::Days         => `time.setUTCDate(time.getUTCDate() + #{offset})`
          Time.Span::Months       => `time.setUTCMonth(time.getUTCMonth() + #{offset})`
          Time.Span::Years        => `time.setUTCFullYear(time.getUTCFullYear() + #{offset})`
        }
      }
      return time;
    })()
    `
  }

  /* Subtracts the delta from the time and returns the result  */
  fun before (delta : Tuple(Number,Time.Span), time : Time) : Time {
    time
    |> Time.shift((amount * -1), unit)
  } where {
    {amount,unit} = delta
  }

  /* Adds the delta from the time and returns the result  */
  fun after (delta : Tuple(Number,Time.Span), time : Time) : Time {
    time
    |> Time.shift(amount, unit)
  } where {
    {amount,unit} = delta
  }
}
```
and this are the tests I've written for it:
```
suite "Time.Span - Initialize spans" {
  test "Define span in milliseconds" {
    Time.milliseconds(1) == {1,Time.Span::Milliseconds}
  }
  test "Define span in seconds" {
    Time.seconds(1) == {1,Time.Span::Seconds}
  }
  test "Define span in minutes" {
    Time.minutes(1) == {1,Time.Span::Minutes}
  }
  test "Define span in hours" {
    Time.hours(1) == {1,Time.Span::Hours}
  }
  test "Define span in days" {
    Time.days(1) == {1,Time.Span::Days}
  }
  test "Define span in weeks" {
    Time.weeks(1) == {7,Time.Span::Days}
  }
  test "Define span in months" {
    Time.months(1) == {1,Time.Span::Months}
  }
  test "Define span in years" {
    Time.years(1) == {1,Time.Span::Years}
  }

}

suite "Time.shift - given some direction, time and a timespan, returns the time corresponding to using the given span and the direction to shift the given time" {
  test "shift 1 millisecond forward" {
    (Time.from(2018, 4, 5)
      |> Time.shift(1, Time.Span::Milliseconds)
      |> Time.toIso()) == "2018-04-05T00:00:00.001Z"
  }
  test "shift 1 millisecond backward" {
    (Time.from(2018, 4, 5)
      |> Time.shift(-1, Time.Span::Milliseconds)
      |> Time.toIso()) == "2018-04-04T23:59:59.999Z"
  }
}

suite "Time.before - given some time and a timespan, returns the time corresponding to subtracting the span from the given time" {
  test "milliseconds before" {
    (Time.from(2018, 4, 5)
    |> Time.before(Time.milliseconds(1))
    |> Time.toIso()) == "2018-04-04T23:59:59.999Z"
  }
  test "seconds before" {
    (Time.from(2018, 4, 5)
    |> Time.before(Time.seconds(1))
    |> Time.toIso()) == "2018-04-04T23:59:59.000Z"
  }
  test "minutes before" {
    (Time.from(2018, 4, 5)
    |> Time.before(Time.minutes(1))
    |> Time.toIso()) == "2018-04-04T23:59:00.000Z"
  }
  test "hours before" {
    (Time.from(2018, 4, 5)
    |> Time.before(Time.hours(1))
    |> Time.toIso()) == "2018-04-04T23:00:00.000Z"
  }
  test "days before" {
    (Time.from(2018, 4, 5)
    |> Time.before(Time.days(1))
    |> Time.toIso()) == "2018-04-04T00:00:00.000Z"
  }
  test "weeks before" {
    (Time.from(2018, 4, 5)
    |> Time.before(Time.weeks(1))
    |> Time.toIso()) == "2018-03-29T00:00:00.000Z"
  }
  test "months before" {
    (Time.from(2018, 4, 5)
    |> Time.before(Time.months(1))
    |> Time.toIso()) == "2018-03-05T00:00:00.000Z"
  }
  test "years before" {
    (Time.from(2018, 4, 5)
    |> Time.before(Time.years(1))
    |> Time.toIso()) == "2017-04-05T00:00:00.000Z"
  }
}
suite "Time.after - given som time and a timespan, returns the time corresponding to adding the span to the given time" {
  test "milliseconds after" {
    (Time.from(2018, 4, 5)
    |> Time.after(Time.milliseconds(1))
    |> Time.toIso()) == "2018-04-05T00:00:00.001Z"
  }
  test "seconds after" {
    (Time.from(2018, 4, 5)
    |> Time.after(Time.seconds(1))
    |> Time.toIso()) == "2018-04-05T00:00:01.000Z"
  }
  test "minutes after" {
    (Time.from(2018, 4, 5)
    |> Time.after(Time.minutes(1))
    |> Time.toIso()) == "2018-04-05T00:01:00.000Z"
  }
  test "hours after" {
    (Time.from(2018, 4, 5)
    |> Time.after(Time.hours(1))
    |> Time.toIso()) == "2018-04-05T01:00:00.000Z"
  }
  test "days after" {
    (Time.from(2018, 4, 5)
    |> Time.after(Time.days(1))
    |> Time.toIso()) == "2018-04-06T00:00:00.000Z"
  }
  test "weeks after" {
    (Time.from(2018, 4, 5)
    |> Time.after(Time.weeks(1))
    |> Time.toIso()) == "2018-04-12T00:00:00.000Z"
  }
  test "months after" {
    (Time.from(2018, 4, 5)
    |> Time.after(Time.months(1))
    |> Time.toIso()) == "2018-05-05T00:00:00.000Z"
  }
  test "years after" {
    (Time.from(2018, 4, 5)
    |> Time.after(Time.years(1))
    |> Time.toIso()) == "2019-04-05T00:00:00.000Z"
  }
}
```

All tests green, but I haven't run mint format or done anything of the sort, though I'll try to remember to do it before making a PR.

I'm still not fully satisfied with the `Tuple` solution, but defining a `record` just for this seems overkill. What's important is to have a consistent data structure which includes both a quantity and its respective "Unit"/"Magnitude" of time (i.e., Seconds, Minutes, Days etc), whether that's achieved by using a Tuple, a Record or by having the value included in the Enum is mostly a matter of how much easier or harder it'll be to maintain that code. Having the Enum hold the value means, if I'm not missing anything, I'd need a whole `case` just to multiply that value by `-1`, and that's not very reasonable. A Record would be easier to work with than a Tuple since it doesn't need destructuring, however that means 4 more lines of code, another possible return type, and all the maintenance burden that would come with it...

For a `record` I'd consider something like:
```
enum Time.Unit {
  Milliseconds
  Seconds
  Minutes
  Hours
  Days
  Months
  Years
}
record Time.Span {
  unit : Time.Unit,
  amount : Number
}
```
which would allow changing, for example, the aforementioned `fun before` to something like:
```
fun before (timespan : Time.Span, time : Time) : Time {
  time
  |> Time.shift((timespan.amount * -1), timespan.unit)
}
```
or, even better, `Time.shift` could also take the record as a parameter, so `Time.before` could simply modify the record when calling `Time.shift`:
```
fun before (timespan : Time.Span, time : Time) : Time {
  time
  |> Time.shift({ timespan | amount = timespan.amount * -1 }))
}
```

Still, as annoying as Destructuring and re-creating the Tuple may be, I'm not sure it is enough to justify all the baggage that will come with adding yet-another-named-thing... I'll sleep on it, maybe I'm missing something, and maybe I'll have some other idea... I'm probably overthinking this anyway, but I'm trying to make it as simple as possible, as I want to minimize any impacts on maintenance. If you guys have any feedback about what you believe would be easier to maintain in the future, or any other comments, I'd love to hear your opinions on this.

**Comment by Sija** [comment]
Closing as the proposal got implemented in #526
